lib/concurrent-merkle-tree/src/merkle_roll.rs

```diff
@@ -196,6 +196,91 @@ impl<const MAX_DEPTH: usize, const MAX_BUFFER_SIZE: usize> MerkleRoll<MAX_DEPTH,
+    pub fn append_subtree(
+        &mut self,
+        subtree_root: Node,
+        subtree_rightmost_leaf: Node,
+        subtree_rightmost_index: u32,
+        subtree_rightmost_proof: Vec<Node>,
+    ) -> Result<Node, CMTError> {
+        check_bounds(MAX_DEPTH, MAX_BUFFER_SIZE);
+        // TODO: consider adding check that the subtree is not empty (but will omit for now)
+        //       note: it will be more time consuming to check that the subtree_root is not a root to an empty tree, because it requires applying O(depth) hashes of trivial nodes
+        if self.rightmost_proof.index >= 1 << MAX_DEPTH {
+            return Err(CMTError::TreeFull);
+        }
+
+        if self.rightmost_proof.index == 0 {
+            // TODO(sorend): Consider if we want to be able to initialize a tree with a subtree append (I think it would be cool, but is a little complex). write an equivallent function for subtree_append: initialize_tree_from_subtree_append
+            // For now
+            return Err(CMTError::CannotInitializeWithSubtreeAppend);
+        }
+
+        // Confirm that subtree_rightmost_proof is valid
+        if recompute(
+            subtree_rightmost_leaf,
+            &subtree_rightmost_proof[..],
+            subtree_rightmost_index - 1,
+        ) != subtree_root
+        {
+            return Err(CMTError::InvalidProof);
+        }
+
+        let leaf = subtree_rightmost_leaf.clone();
+        let intersection = self.rightmost_proof.index.trailing_zeros() as usize;
+
+        // @dev: At any given time (other than initialization), there is only one valid size of subtree that can be appended
+        if subtree_rightmost_proof.len() != intersection {
+            return Err(CMTError::SubtreeInvalidSize);
+        }
+
+        let mut change_list = [EMPTY; MAX_DEPTH];
+        let mut intersection_node = self.rightmost_proof.leaf;
+
+        // This will be mutated into the new root after the append by gradually hashing this node with the RMP to the subtree, then the critical node, and then the rest of the RMP to this tree.
+        let mut node = subtree_rightmost_leaf;
+
+        for i in 0..MAX_DEPTH {
+            change_list[i] = node;
+            if i < intersection {
+                // Compute proof to the appended node from empty nodes
+                hash_to_parent(
+                    &mut intersection_node,
+                    &self.rightmost_proof.proof[i],
+                    ((self.rightmost_proof.index - 1) >> i) & 1 == 0,
+                );
+                hash_to_parent(
+                    &mut node,
+                    &subtree_rightmost_proof[i],
+                    ((subtree_rightmost_index - 1) >> i) & 1 == 0,
+                );
+                self.rightmost_proof.proof[i] = subtree_rightmost_proof[i];
+            } else if i == intersection {
+                // Compute where the new node intersects the main tree
+                hash_to_parent(&mut node, &intersection_node, false);
```

Comment:
To check my understanding, at this point `node == root(subtree)`. For extra precaution, can we assert that here?